client/my-sites/site-settings/wrap-settings-form.jsx

```diff
@@ -0,0 +1,179 @@
+import { saveSiteSettings } from 'state/site-settings/actions';
+import { removeNotice, successNotice, errorNotice } from 'state/notices/actions';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import QuerySiteSettings from 'components/data/query-site-settings';
+
+const wrapSettingsForm = getFormSettings => SettingsForm => {
+	class WrappedSettingsForm extends Component {
+		state = {
+			uniqueEvents: {}
+		};
+
+		componentWillMount() {
+			this.props.replaceFields( getFormSettings( this.props.settings ) );
+		}
+
+		componentWillReceiveProps( nextProps ) {
+			if ( nextProps.siteId !== this.props.siteId ) {
+				nextProps.clearDirtyFields();
+			}
+
+			if ( nextProps.settings !== this.props.settings ) {
+				let newState = getFormSettings( nextProps.settings );
+				//If we have any fields that the user has updated,
+				//do not wipe out those fields from the poll update.
+				newState = omit( newState, nextProps.dirtyFields );
+				nextProps.replaceFields( newState );
+			}
+
+			if (
+				this.props.isSavingSettings &&
+				! nextProps.isSavingSettings
+			) {
+				if ( nextProps.isSaveRequestSuccessful ) {
+					nextProps.successNotice( nextProps.translate( 'Settings saved!' ), { id: 'site-settings-save' } );
+					nextProps.clearDirtyFields();
+					nextProps.markSaved();
+				} else {
+					let text;
+					switch ( nextProps.saveRequestError.error ) {
+						case 'invalid_ip':
+							text = nextProps.translate( 'One of your IP Addresses was invalid. Please try again.' );
+							break;
+						default:
+							text = nextProps.translate( 'There was a problem saving your changes. Please try again.' );
+					}
+					nextProps.errorNotice( text, { id: 'site-settings-save' } );
+				}
+			}
+		}
+
+		// Some Utils
+		handleSubmitForm = event => {
+			if ( ! event.isDefaultPrevented() && event.nativeEvent ) {
+				event.preventDefault();
+			}
+
+			this.submitForm();
+			this.props.trackEvent( 'Clicked Save Settings Button' );
+		};
+
+		submitForm = () => {
+			const { fields, site } = this.props;
+			this.props.removeNotice( 'site-settings-save' );
+			this.props.saveSiteSettings( site.ID, fields );
+		};
+
+		handleRadio = event => {
+			const currentTargetName = event.currentTarget.name,
+				currentTargetValue = event.currentTarget.value;
+
+			this.props.updateFields( { [ currentTargetName ]: currentTargetValue } );
+		};
+
+		handleToggle = name => () => {
+			this.props.trackEvent( `Toggled ${ name }` );
+			this.props.updateFields( { [ name ]: ! this.props.fields[ name ] } );
+		};
+
+		onChangeField = field => event => {
+			const { updateFields } = this.props;
+			updateFields( {
+				[ field ]: event.target.value
+			} );
+		};
+
+		uniqueEventTracker = message => () => {
+			if ( this.state.uniqueEvents[ message ] ) {
+				return;
+			}
+			const uniqueEvents = {
+				...this.state.uniqueEvents,
+				[ message ]: true,
+			};
+			this.setState( { uniqueEvents } );
+			this.state.uniqueEvents[ message ] = true;
```

Comment:
We don't want to assign into state directly like this, and at the same time, I don't think it's necessary, as I believe your new `uniqueEvents` object created above should account for this already?